File: 3737. count-subarrays-with-majority-element-i.py
from typing import List

class Solution:
    def countMajoritySubarrays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        ans = 0

        for l in range(n):
            target_count = 0

            for r in range(l, n):
                if nums[r] == target:
                    target_count += 1

                length = r - l + 1

                if target_count > length // 2:
                    ans += 1

        return ans


if __name__ == "__main__":
    sol = Solution()

    print(sol.countMajoritySubarrays([1, 2, 2, 3], 2))  # Expected: 5
    print(sol.countMajoritySubarrays([1, 1, 1, 1], 1))  # Expected: 10
    print(sol.countMajoritySubarrays([1, 2, 3], 4))     # Expected: 0

File: 3737. count-subarrays-with-majority-element-i.java
class Solution {
    public int countMajoritySubarrays(int[] nums, int target) {
        int n = nums.length;
        int balance = n + 1;

        int[] freq = new int[2 * n + 2];
        int[] pref = new int[2 * n + 2];

        freq[balance] = 1;
        pref[balance] = 1;

        int answer = 0;

        for (int num : nums) {
            if (num == target) {
                balance++;
            } else {
                balance--;
            }
            freq[balance]++;

            pref[balance] = pref[balance - 1] + freq[balance];

            answer += pref[balance - 1];
        }

        return answer;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        System.out.println(sol.countMajoritySubarrays(
            new int[]{1, 2, 2, 3}, 2
        )); // Expected: 5

        System.out.println(sol.countMajoritySubarrays(
            new int[]{1, 1, 1, 1}, 1
        )); // Expected: 10

        System.out.println(sol.countMajoritySubarrays(
            new int[]{1, 2, 3}, 4
        )); // Expected: 0
    }
}